### ISSUE TITLE
subunit reporter does not handle import errors

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4496]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#4496|
|Type|defect|
|Created|2010-06-07 23:55:41Z|
|Branch|https://github.com/twisted/twisted/tree/subunit-import-error-4496|

Create a file, foo.py, with these contents:
```python
import doesntexist
```

Run `trial foo`, and you'll see a nice error about not being able to import `doesntexist`.

Run `trial --reporter=subunit` and you'll see:
```
error: foo [
Traceback (most recent call last):
  File "/home/jml/src/Twisted/trunk/bin/trial", line 22, in <module>
    run()
  File "/home/jml/src/Twisted/trunk/twisted/scripts/trial.py", line 363, in run
    test_result = trialRunner.run(suite)
  File "/home/jml/src/Twisted/trunk/twisted/trial/runner.py", line 835, in run
    return self._runWithoutDecoration(test)
  File "/home/jml/src/Twisted/trunk/twisted/trial/runner.py", line 864, in _runWithoutDecoration
    run()
  File "/home/jml/src/Twisted/trunk/twisted/trial/runner.py", line 859, in <lambda>
    run = lambda: suite.run(result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/runner.py", line 298, in run
    TestSuite.run(self, result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/unittest.py", line 1390, in run
    test(result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/unittest.py", line 1378, in __call__
    return self.run(result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/runner.py", line 174, in run
    super(LoggedSuite, self).run(result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/unittest.py", line 1390, in run
    test(result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/unittest.py", line 1378, in __call__
    return self.run(result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/runner.py", line 147, in run
    test(result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/runner.py", line 388, in __call__
    return self.run(result)
  File "/home/jml/src/Twisted/trunk/twisted/trial/runner.py", line 384, in run
    result.addError(self, self.error)
  File "/home/jml/src/Twisted/trunk/twisted/trial/reporter.py", line 1013, in addError
    test, util.excInfoOrFailureToExcInfo(err))
  File "/usr/lib/python2.6/dist-packages/subunit/__init__.py", line 569, in addError
    self._addOutcome("error", test, error=error, details=details)
  File "/usr/lib/python2.6/dist-packages/subunit/__init__.py", line 621, in _addOutcome
    for line in self._exc_info_to_string(error, test).splitlines():
  File "/usr/lib/python2.6/unittest.py", line 156, in _exc_info_to_string
    if exctype is test.failureException:
AttributeError: 'ErrorHolder' object has no attribute 'failureException'
```

The subunit error should not refer to `ErrorHolder` and instead have the import error that is the actual cause of the failure.

<details><summary>Searchable metadata</summary>

```
trac-id__4496 4496
type__defect defect
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_subunit_import_error_4496 branches/subunit-import-error-4496
branch_author__jml jml
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1275954941000000 1275954941000000
changetime__1276517442000000 1276517442000000
version__None None
owner__ 
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

(In [29248]) Branching to 'subunit-import-error-4496'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml removed owner|
|-|-|

As lifeless astutely observed, this bug is actually because `ErrorHolder` doesn't behave enough like a `TestCase` for Python's testing library.

To remedy this, I've added one test that corresponds to the bug (`test_loadTimeErrors`) as well as some direct tests against `ErrorHolder` and its base class `TestHolder`.

The direct tests involve actually running the `Holder` twins against a standard `TestResult` object and checking that the result object thinks it ran a test, and has some idea of the result of that test.

I've also beefed up the interface verification test to actually verify the interface. This forced a bunch of stuff from `ErrorHolder` up onto `TestHolder`. It's worth noting that if the interface verification test were like this in the beginning, the bug would never have happened.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess set owner to @jml|
|-|-|

I could only force builds on half the machines for some reason, but those results are [here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/subunit-import-error-4496) and are fine.

test_runner.py needs a copyright bump, and in runner.py the two changed run()s could use markup for the ```result``` parameter. Other than that, looks good and ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml set status to `closed`|
|-|-|

(In [29304]) Merge subunit-import-error-4496

 * Author: jml
 * Reviewer: jesstess
 * Fixes [#4496](https://github.com/twisted/twisted/issues/4496)

The subunit reporter used to crash and die when it encountered an import error.
Now, it reports it properly.

This was due to the `ErrorHolder` and `TestHolder` objects not looking enough
like `TestCase`s.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
